### PR TITLE
Wildcard index type support for associative arrays

### PR DESCRIPTION
**Pull request by kozdra**
Associative arrays that specify a wildcard index type may be indexed by
integral expressions of any size, with leading zeros removed
automatically.  A natural representation for such expressions is a
string, especially that the standard explicitly specifies automatic
casts from string indices to bit vectors of equivalent size.
The automatic cast part is done implicitly by the existing type system.

A simpler way to just make this work would be to convert wildcard index
type to a string type directly in the parser code, but several new AST
classes are needed to make sure illegal method calls are detected.
The verilated data structure implementation is reused, because there is
no need for differentiating the behavior on C++ side.

### CODE REVIEW

**Comment by kbieganski** [comment]
What happens if I do:

```systemverilog
a[1'b0] = 1;
a[8'b0] = 2;
a["0"] = 3;
```

Will I have 3 different entries? What should be the result?

**Comment by kozdra** [comment]
It will have one entry, as all accesses will be cast to the string "0".

File: src/V3Ast.h [hunk 1]
```diff
@@ -1719,7 +1719,7 @@ class AstNode VL_NOT_FINAL {
     // Data type locators
     AstNodeDType* findBitDType() { return findBasicDType(VBasicDTypeKwd::LOGIC); }
     AstNodeDType* findDoubleDType() { return findBasicDType(VBasicDTypeKwd::DOUBLE); }
-    AstNodeDType* findStringDType() { return findBasicDType(VBasicDTypeKwd::STRING); }
+    AstNodeDType* findStringDType() const { return findBasicDType(VBasicDTypeKwd::STRING); }
```
Comment (kbieganski) [inline src/V3Ast.h:1722]: Do we need this change in this PR? If not, let's just leave it non-const. This fix should be a separate PR.

File: src/V3AstNodes.h [hunk 2]
```diff
@@ -1686,6 +1749,44 @@ class AstAssocSel final : public AstNodeSel {
     virtual int instrCount() const override { return widthInstrs(); }
 };
 
+class AstWildcardSel final : public AstNodeSel {
```
Comment (kbieganski) [inline src/V3AstNodes.h:1752]: As I said before, I'm not sure we need this, but we'll see what the Verilator people say about the whole thing. Let's keep it for now.

File: src/V3Width.cpp [hunk 3]
```diff
@@ -981,6 +981,27 @@ class WidthVisitor final : public VNVisitor {
...
+            const AstWildcardArrayDType* const adtypep = VN_CAST(fromDtp, WildcardArrayDType);
+            if (!adtypep) {
+                UINFO(1, "    Related dtype: " << fromDtp << endl);
+                nodep->v3fatalSrc("Wildcard array reference is not to wildcard array");
+            }
+            const AstBasicDType* const basicp = nodep->bitp()->dtypep()->skipRefp()->basicp();
+            if (!basicp
+                || (basicp->keyword() != VBasicDTypeKwd::STRING
+                    && !basicp->keyword().isIntNumeric())) {
+                nodep->v3error("Wildcard index must be integral (IEEE 1800-2017 7.8.1)");
```
Comment (kbieganski) [inline src/V3Width.cpp:997]: Let's add a test for this error. See tests named like `t_*_bad.pl`.

Similarly, test all the other errors you added in this PR, like the invalid method errors. (user-facing ones, not asserts)

**Review by kbieganski — CHANGES_REQUESTED** [review]
Overall I like it, nice job. Just a few small things.

File: src/V3Ast.h (hunk 1 above)
Comment (kozdra) [reply src/V3Ast.h:1722]: This is necessary, because it is used in `const` context 5 times (every time we call `findStringDType` here).  In hindsight, it might also be necessary to apply the "must be integral" check every time we call `findStringDType`, because it is related to typing the array index.

File: src/V3Width.cpp (hunk 3 above)
Comment (kozdra) [reply src/V3Width.cpp:997]: Done :smiley:

**Review by kbieganski — APPROVED** [review]
Ok, LGTM
